pkg/cstorvolumereplica/v1alpha1/kubernetes.go

```diff
@@ -0,0 +1,97 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	kclient "github.com/openebs/maya/pkg/client/k8s/v1alpha1"
+)
+
+// getClientsetFn is a typed function that
+// abstracts fetching of internal clientset
+type getClientsetFn func() (clientset *clientset.Clientset, err error)
+
+// listFn is a typed function that abstracts
+// listing of cstor volume replica instances
+type listFn func(cli *clientset.Clientset, namespace string, opts metav1.ListOptions) (*apis.CStorVolumeReplicaList, error)
+
+// kubeclient enables kubernetes API operations
+// on cstor volume replica instance
+type kubeclient struct {
+	// clientset refers to cstor volume replica's
+	// clientset that will be responsible to
+	// make kubernetes API calls
+	clientset clientset.Interface
+
+	// functions useful during mocking
+	getClientset getClientsetFn
+	list         listFn
+}
+
+// kubeclientBuildOption defines the abstraction
+// to build a kubeclient instance
+type kubeclientBuildOption func(*kubeclient)
+
+// withDefaults sets the default options
+// of kubeclient instance
+func (k *kubeclient) withDefaults() {
+	if k.getClientset == nil {
+		k.getClientset = func() (clients *clientset.Clientset, err error) {
+			config, err := kclient.Config().Get()
+			if err != nil {
+				return nil, err
+			}
+
+			return clientset.NewForConfig(config)
+		}
+	}
+	if k.list == nil {
+		k.list = func(cli *clientset.Clientset, namespace string, opts metav1.ListOptions) (*apis.CStorVolumeReplicaList, error) {
+			return cli.OpenebsV1alpha1().CStorVolumeReplicas(namespace).List(opts)
+		}
+	}
+}
+
+// WithKubeClient sets the kubernetes client against
+// the kubeclient instance
+func WithKubeClient(c clientset.Interface) kubeclientBuildOption {
+	return func(k *kubeclient) {
+		k.clientset = c
+	}
+}
+
+// KubeClient returns a new instance of kubeclient meant for
+// cstor volume replica operations
+func KubeClient(opts ...kubeclientBuildOption) *kubeclient {
+	k := &kubeclient{}
+	for _, o := range opts {
+		o(k)
+	}
+	k.withDefaults()
+	return k
+}
+
+// getClientOrCached returns either a new instance
+// of kubernetes client or its cached copy
+func (k *kubeclient) getClientOrCached() (clientset.Interface, error) {
+	if k.clientset != nil {
+		return k.clientset, nil
+	}
+	c, err := k.getClientset()
+	if err != nil {
+		return nil, err
+	}
+	k.clientset = c
+	return k.clientset, nil
+}
+
+// List returns a list of cstor volume replica
+// instances present in kubernetes cluster
+func (k *kubeclient) List(namespace string, opts metav1.ListOptions) (*apis.CStorVolumeReplicaList, error) {
+	cli, err := k.getClientOrCached()
+	if err != nil {
+		return nil, err
+	}
+	return k.list(cli.(*clientset.Clientset), namespace, opts)
```

Comment:
This type casting may not be required if `getClientOrCached` signature is modified as per earlier comment